### PR TITLE
Add support for bn254

### PR DESCRIPTION
**Pull request by Schaeff**
- move methods to a trait
- implement trait for goldilocks and bn254
- change AbstractNumberType to BigUint
- use `ark_ff::BigInteger` for bitwise operations (in particular in bit constraints)

### CODE REVIEW

**Comment by leonardoalt** [comment]
In the riscv zkVM there are lots of assumptions on GL

File: number/src/traits.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+use num_traits::Zero;
+use std::{
+    fmt,
+    ops::{Add, AddAssign, Div, Mul, Neg, Sub},
+};
+
+use crate::{AbstractNumberType, DegreeType};
+pub trait FieldElementTrait:
```
Comment (leonardoalt) [inline number/src/traits.rs:8]: Why do we need a new trait instead of just using something from ark?

File: number/src/goldilocks.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+use ark_ff::{Fp64, MontBackend, MontConfig};
+
+#[derive(MontConfig)]
+#[modulus = "18446744069414584321"]
+#[generator = "7"]
+pub struct GoldilocksBaseFieldConfig;
+pub type GoldilocksBaseField = Fp64<MontBackend<GoldilocksBaseFieldConfig, 1>>;
```
Comment (leonardoalt) [inline number/src/goldilocks.rs:7]: Does ark not have GL?

File: number/src/traits.rs (hunk 1 above)
Comment (Schaeff) [reply number/src/traits.rs:8]: We currently already wrap ark. This PR just moves the methods to a trait. If we wanna use ark directly we need to change the call sites of these methods to using ark only. Also (less important) this lets us support something else than ark if needed by just implementing the trait.

File: number/src/goldilocks.rs (hunk 2 above)
Comment (Schaeff) [reply number/src/goldilocks.rs:7]: Nein https://github.com/arkworks-rs/curves
Also note that even the curves ark "has" just use this macro under the hood for the field.

**Comment by Schaeff** [comment]
> In the riscv zkVM there are lots of assumptions on GL

Once we expose both GL and Bn we can make the riscv crate use `GoldilocksField` directly instead of being generic over `F: FieldElementTrait`

File: number/src/traits.rs (hunk 1 above)
Comment (leonardoalt) [reply number/src/traits.rs:8]: ok

**Comment by leonardoalt** [comment]
> > In the riscv zkVM there are lots of assumptions on GL
> 
> Once we expose both GL and Bn we can make the riscv crate use `GoldilocksField` directly instead of being generic over `F: FieldElementTrait`

So what's missing currently?

**Comment by Schaeff** [comment]
> > > In the riscv zkVM there are lots of assumptions on GL
> > 
> > 
> > Once we expose both GL and Bn we can make the riscv crate use `GoldilocksField` directly instead of being generic over `F: FieldElementTrait`
> 
> So what's missing currently?

In this PR? Fixing the tests and figuring out what to use for bigint. This PR does not make the whole codebase generic, it only introduces the trait, everything still uses GL.

File: number/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            type Output = $name;
+
+            fn add(self, rhs: Self) -> Self::Output {
+                $name {
+                    value: self.value + rhs.value,
+                }
+            }
+        }
+
+        impl<'a> std::ops::Add<$name> for &'a $name {
```
Comment (chriseth) [inline number/src/macros.rs:95]: Are the implementations for the references not automatic?

File: executor/src/witgen/bit_constraints.rs [hunk 4]
```diff
@@ -52,7 +56,9 @@ impl BitConstraint {
     /// The bit constraint of an integer multiple of an expression.
     /// TODO this assumes goldilocks
     pub fn multiple(&self, factor: FieldElement) -> Option<BitConstraint> {
-        if factor.to_integer() * self.mask >= FieldElement::modulus() {
+        if factor.to_arbitrary_integer() * self.mask.into_biguint()
```
Comment (Schaeff) [inline executor/src/witgen/bit_constraints.rs:59]: @chriseth could you explain this line? I'm a bit surprised to see a multiplication by a mask

File: number/src/macros.rs (hunk 3 above)
Comment (Schaeff) [reply number/src/macros.rs]: I removed these as it's not needed now that the underlying integer is Copy

File: executor/src/witgen/bit_constraints.rs (hunk 4 above)
Comment (chriseth) [reply executor/src/witgen/bit_constraints.rs:59]: Hm yeah, interesting. This of course only works if `factor` is a power of two. Is that checked outside?

Comment (leonardoalt) [reply executor/src/witgen/bit_constraints.rs:59]: So what's the conclusion here?

Comment (Schaeff) [reply executor/src/witgen/bit_constraints.rs:57]: It's not clear to me that `factor` is always a power of two, as it is a coefficient in an arbitrary `AffineExpression`.

**Comment by Schaeff** [comment]
There are more additions to these traits in #256 but maybe it's worth merging this first.

File: executor/src/witgen/affine_expression.rs [hunk 5]
```diff
@@ -235,27 +235,31 @@ where
...
                 i,
-                Constraint::Assignment(((offset & mask) / coeff.to_integer()).into()),
+                Constraint::Assignment(
+                    ((offset & *mask).into_biguint() / coeff.to_arbitrary_integer())
+                        .try_into()
+                        .unwrap(),
+                ),
             )]));
-            offset &= !mask;
+            offset &= !*mask
```
Comment (leonardoalt) [inline executor/src/witgen/affine_expression.rs:259]: Why is this returned now?

File: number/src/goldilocks.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        let _ = GoldilocksField::from(1).to_arbitrary_integer()
+            / GoldilocksField::from(0).to_arbitrary_integer();
+    }
+
+    #[test]
+    #[should_panic]
+    fn div_by_zero() {
+        let _ = GoldilocksField::from(1) / GoldilocksField::from(0);
+    }
+}
```
Comment (leonardoalt) [inline number/src/goldilocks.rs:27]: Maybe add similar tests to BN?

File: pil_analyzer/src/pil_analyzer.rs [hunk 7]
```diff
@@ -653,12 +653,27 @@ impl PILContext {
...
+                .try_into()
+                .unwrap(),
+                BinaryOperator::BinaryXor => (left.to_arbitrary_integer()
+                    ^ right.to_arbitrary_integer())
+                .try_into()
+                .unwrap(),
+                BinaryOperator::BinaryOr => (left.to_arbitrary_integer()
+                    | right.to_arbitrary_integer())
+                .try_into()
+                .unwrap(),
```
Comment (leonardoalt) [inline pil_analyzer/src/pil_analyzer.rs:670]: Indentation looks weird

File: executor/src/witgen/affine_expression.rs (hunk 5 above)
Comment (Schaeff) [reply executor/src/witgen/affine_expression.rs:259]: Interesting, `&=` returns nothing so these are equivalent I think.

File: number/src/traits.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    type Integer: BigInt;
+
+    fn to_degree(&self) -> DegreeType;
+
+    fn to_integer(&self) -> Self::Integer;
+
+    fn to_arbitrary_integer(&self) -> AbstractNumberType {
+        self.to_integer().into_biguint()
+    }
+
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply number/src/traits.rs:46]: I should add docs on this. Basically `AbstractNumberType = BigUint` is vector based arbitrary size integer, slower at runtime. `FieldElementTrait::Integer: BigInt` is static array integer, used for fast operations on masks.

File: pil_analyzer/src/pil_analyzer.rs (hunk 7 above)
Comment (Schaeff) [reply pil_analyzer/src/pil_analyzer.rs:670]: fmt 🤷 will see what I can do

File: executor/src/witgen/affine_expression.rs [hunk 9]
```diff
@@ -235,27 +235,31 @@ where
...
                 ));
             } else {
-                covered_bits |= mask;
+                covered_bits |= *mask;
             }
             assignments.combine(EvalValue::complete(vec![(
                 i,
-                Constraint::Assignment(((offset & mask) / coeff.to_integer()).into()),
+                Constraint::Assignment(
+                    ((offset & *mask).into_biguint() / coeff.to_arbitrary_integer())
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply executor/src/witgen/affine_expression.rs:254]: The former is for `BigInt`, the latter is for `FieldElement`

File: executor/src/witgen/affine_expression.rs (hunk 5 above)
Comment (leonardoalt) [reply executor/src/witgen/affine_expression.rs:259]: Then maybe let's keep `;`?

File: pil_analyzer/src/pil_analyzer.rs (hunk 7 above)
Comment (leonardoalt) [reply pil_analyzer/src/pil_analyzer.rs:670]: ah, weird... some items are similar to this but have the "correct" indentation

Comment (Schaeff) [reply pil_analyzer/src/pil_analyzer.rs:670]: This can actually use fixed width integers

File: executor/src/witgen/bit_constraints.rs (hunk 4 above)
Comment (leonardoalt) [reply executor/src/witgen/bit_constraints.rs]: I would approve but don't know what needs to be done here

Comment (Schaeff) [reply executor/src/witgen/bit_constraints.rs]: I moved this to #270

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**